gk/fib.c

```diff
@@ -1696,6 +1723,170 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+		if (ret < 0) {
+			ret = snprintf(err_buf, err_buf_len,
+				"Cannot convert gateway IP %s to bytes\n",
+				gw_ip);
+			RTE_VERIFY(ret > 0 && ret < (int)err_buf_len);
+			return -1;
+		}
+
+		/* Pop this Grantor/gateway IP pair and strings off stack. */
+		lua_pop(l, 3);
+	}
+
+	return 0;
+}
+
+static void
+add_grantor_entry_lb_verify_params(lua_State *l, const char **prefix,
+	lua_Integer *tbl_size, struct gk_config **gk_conf)
+{
+	uint32_t ctypeid;
+	uint32_t correct_ctypeid_gk_config = luaL_get_ctypeid(l,
+		CTYPE_STRUCT_GK_CONFIG_PTR);
+	void *cdata;
+	size_t len;
+
+	if (lua_gettop(l) != 3) {
+		luaL_error(l,
+			"%s expected three arguments, however it received %d arguments\n",
+		      __func__, lua_gettop(l));
+	}
+
+	/* First argument must be a prefix string. */
+	*prefix = lua_tolstring(l, 1, &len);
+	if (*prefix == NULL || len == 0)
+		luaL_error(l, "Could not read prefix for adding load balanced Grantor set\n");
+
+	/* Second argument must be a table. */
+	luaL_checktype(l, 2, LUA_TTABLE);
+	*tbl_size = lua_objlen(l, 2);
+	if (*tbl_size <= 0)
+		luaL_error(l, "Table must have a positive number of Grantor entries\n");
+
+	/* Third argument must be of type CTYPE_STRUCT_GK_CONFIG_PTR. */
+	cdata = luaL_checkcdata(l, 3,
+		&ctypeid, CTYPE_STRUCT_GK_CONFIG_PTR);
+	if (ctypeid != correct_ctypeid_gk_config) {
+		luaL_error(l, "%s expected '%s' as the third argument\n",
+			__func__, CTYPE_STRUCT_GK_CONFIG_PTR);
+	}
+	*gk_conf = *(struct gk_config **)cdata;
+}
+
+int
+l_add_grantor_entry_lb(lua_State *l)
+{
+	const char *prefix;
+	struct ip_prefix prefix_info;
+	lua_Integer tbl_size;
+	struct gk_config *gk_conf;
+	struct ipaddr *gt_addrs;
+	struct ipaddr *gw_addrs;
+	char err_buf[256];
+	const char *err_msg = NULL;
+	int ret;
+
+	/* Verify presence and types of parameters and read them in. */
+	add_grantor_entry_lb_verify_params(l, &prefix,
+		&tbl_size, &gk_conf);
+
+	gt_addrs = rte_malloc("grantor_ips",
+		tbl_size * sizeof(*gt_addrs), 0);
+	if (gt_addrs == NULL) {
+		err_msg = "Not enough memory to add Grantor entries\n";
+		goto out;
+	}
+
+	gw_addrs = rte_malloc("gateway_ips",
+		tbl_size * sizeof(*gw_addrs), 0);
+	if (gw_addrs == NULL) {
+		err_msg = "Not enough memory to add gateway entries\n";
+		goto free_gt;
+	}
+
+	/* Read in pairs of Grantor/gateway addresses from table. */
+	if (read_grantor_entries(l, gt_addrs, gw_addrs,
+			err_buf, sizeof(err_buf)) < 0) {
+		err_msg = err_buf;
+		goto free_gw;
+	}
+
+	/* Set up prefix info. */
+	prefix_info.str = prefix;
+	prefix_info.len = parse_ip_prefix(prefix, &prefix_info.addr);
+
+	ret = add_fib_entry_numerical(&prefix_info,
+		gt_addrs, gw_addrs, tbl_size, GK_FWD_GRANTOR,
+		RTPROT_STATIC, gk_conf);
+	if (ret < 0)
+		err_msg = "Could not add FIB entry; check Gatekeeper log\n";
+
+free_gw:
+	rte_free(gw_addrs);
+free_gt:
+	rte_free(gt_addrs);
+out:
+	if (err_msg != NULL)
+		luaL_error(l, err_msg);
+
+	lua_pushnumber(l, 0);
+	return 1;
```

Comment:
We are always returning zero here, don you mean to return `ret` instead?

Reply:
Since all error conditions use `luaL_error()` to raise an error handled by Lua, there is no condition under which this function returns a non-zero value.

Reply:
So let's return nothing instead.